### ISSUE TITLE
Give AudioStreamPlayer a property or method to provide similar functionality to "PlayOneShot()" from Unity's AudioSources.

### ISSUE THREAD

**Issue by amarraff** [godotengine/godot-proposals#8903]
### Describe the project you are working on

A 3D game with sounds that play as the player navigates the UI. These sounds can be played rapidly, depending on how fast the user navigates the UI.

### Describe the problem or limitation you are having in your project

When play() is called on an AudioStreamPlayer, it will play the assigned AudioStream. However, if you call play() again while the AudioStream is still playing, it will get cut off, and the new AudioStream (whether the same audio file, or a new one) will begin playing instead. The desired behavior would be similar to "PlayOneShot()" from Unity's AudioSources. Documentation for this method here: https://docs.unity3d.com/ScriptReference/AudioSource.PlayOneShot.html  

I am aware that there is a type of AudioStream called AudioStreamPolyphonic, but I do not understand how to implement it. I have searched the internet, and tried messing with it myself, but have been unsuccessful in getting it to work with the desired behavior I described. If it is simple to do, I am sorry, I could not figure it out. I also tried increasing the value of "Max Polyphony" in the AudioStreamPlayer, but this did not work as expected either.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a property or method to AudioStreamPlayer, AudioStreamPlayer2D, and AudioStreamPlayer3D that allows playing multiple AudioStreams without successive plays interrupting the previous ones. This would allow playing UI sounds, or other common game sounds like weapon bullets, rapidly without the AudioStreams cutting each other off; the full audio file would play each time, even when successive AudioStreams are played.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

- **If implemented as a property:**

Give AudioStreamPlayers a boolean property with a name like "Play Polyphonic" which, when true, would enable this functionality for every AudioStream the player plays.

- **If implemented as a method:**

Add a method similar to Unity's "PlayOneShot()" as a partner function to AudioStreamPlayer's "play()". For example, "_play_polyphonic()". The user could pass a desired AudioStream though this function to achieve the desired functionality. 
For example: my_audio_stream_player._play_polyphonic(my_stream)

- **Both of these implementations should result in the same functionality:** 

1.) Tell the AudioStreamPlayer to create a new AudioStreamPlayer instance which inherits all of its values (including transform for 2D & 3D AudioStreamPlayers), and add it to the tree as a child of the original AudioStreamPlayer.
2.) Play the provided AudioStream with it. 
3.) Call queue_free() on the instanced player when its "finished" signal is fired. 

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I believe this enhancement would be used quite often. However, this behavior can be done manually via code. In my example below, "_play_polyphonic()" is inside an AutoLoad script called AudioStreamFunctions. So, anything can call AudioStreamFunctions._play_polyphonic(AudioStream) to achieve this desired result.

```
extends Node

func _play_polyphonic(stream:AudioStream):
	var new_player:AudioStreamPlayer = AudioStreamPlayer.new()
	get_tree().root.add_child(new_player)
	new_player.stream = stream
	new_player.play()
	await new_player.finished
	new_player.queue_free()
```

However, this is not particularly ideal if the user wants each AudioStream to follow specific settings for volume, pitch, bus, etc. on each play, because it creates a new blank AudioStreamPlayer each play. So, the user must do something like this:

- Create an AudioStreamPlayer in the node hierarchy, set desired property values, assign it to an @export var, then instance it, play a stream from that instance, and free it.

- Create a function to quickly assign desired settings to each new AudioStreamPlayer immediately after it is instanced.

- Pass through an AudioStreamPlayer to instantiate along with the AudioStream. 

Again, this is just extra work on the user's end. It would be great to have the feature built-in to AudioStreamPlayer nodes, like Unity has with their AudioSources.

### Is there a reason why this should be core and not an add-on in the asset library?

It is a quality of life addition to the existing AudioStreamPlayer nodes. I don't think it would make sense to make it an add-on.

**Comment by AThousandShips** [comment]
See also:
* https://github.com/godotengine/godot-proposals/issues/7350

**Comment by amarraff** [comment]
> See also:
> 
>     * [allow and add features to play sound files directly in code without instancing audio stream player nodes. #7350](https://github.com/godotengine/godot-proposals/issues/7350)

This seems to be a different issue. That user is talking about playing sound files inside of code without using an AudioStreamPlayer at all. I'm talking about using one AudioStreamPlayer to play multiple uninterrupted AudioStreams.

**Comment by AThousandShips** [comment]
I didn't say the same thing 🙂 it's related but not identical, it's about playing without a permanent player

**Comment by amarraff** [comment]
> I didn't say the same thing 🙂 it's related but not identical, it's about playing without a permanent player

No worries, I couldn't tell what you were implying by suggesting I look at it.

**Comment by AThousandShips** [comment]
Now this would be an entirely unique feature, nowhere else do we have this feature of automatically spawning and removing nodes, and the workaround is very trivial

This solves a very specific issue, not a general one, and is better solved IMO by user code, adapted to the specific use case, this is not something I've ever needed myself, or seen anyone else ask for or need, so I don't feel this is a widely needed feature

So I'd say this belongs in a plugin or as sample code, as opposed to in the engine, as I don't see the real need for it, or the need being common

**Comment by amarraff** [comment]
I respectfully disagree, this is an issue many users will run into. It may be trivial for you, but it took me a while to figure out the right way to achieve this, and I’ve been coding for years. A beginner will struggle more. 

I did mention the polyphonic audio stream in my post,I don’t understand how to use it. Better documentation on that page with syntax examples would be great. 

I also mentioned that the solution I provided is not ideal. It would be better to have the AudioStreamPlayer itself to handle this functionality. It doesn’t have to be in the specific way I outlined, but I imagine any implementation of it would probably do something similar,

Is there a best practice way to do this in Godot? How would you handle, for example, a machine gun sound, so that the sound of each bullet firing does not cut off the previous one?

**Comment by KoBeWi** [comment]
AudioStreamPlayers have `max_polyphony` property that does what you want.

**Comment by AThousandShips** [comment]
Have you tried asking on the other channels how to use that class? It should solve your problems without adding new unusual functionality, this method isn't how things are done in Godot, we don't spawn and then delete nodes otherwise, please first confirm you can't achieve this with the polyphonic stream

As I said documentation for that class will likely solve this, so that's the proper solution, *unless* it won't solve your problem, please try other solutions first

**Comment by amarraff** [comment]
@AThousandShips 
- I asked in the Godot discord but didn't manage to get a reply. I also tried searching the net for an answer but didn't find anything in forums on YouTube. The docs also just confused me, I don't understand how to implement it. I already mentioned that in my original post. I also already mentioned that my method of spawning and deleting is simply because I don't know a better way to do it, and that it does not have to be accomplished that specific way. Just the property or the method for the user.

- As I already said in my original post, I tried getting the polyphonic audio stream class working, but could not figure it out. I did try other solutions, like the code I provided in my post...

- You'd use a sound clip? Like a pre-made loop of the bullets firing? That doesn't sound like a good way to do it to me. How would you stop the sound clip at the right point to ensure the sound of the bullets firing sounds good? Bullets have their initial firing sound, followed by an echo usually. That full sound needs to play. If you pre-record a sound file of bullets overlapping each other, you won't be able to hear the sound of the last bullet finish without another bullet overlapped on top of it. 

- How would animation achieve this functionality?

**Comment by amarraff** [comment]
@Mickeon 
As I said in my original post, I already tried raising that value. It does not work. I was hoping it would be that easy. I am aware that there is an AudioStreamPolyphonic class, but I do not understand how to use it. This confusing led me to suggest this proposal, so make implementing this feature easier for all users.

**Comment by AThousandShips** [comment]
Did you read the page I linked? It has audio support which does polyphonic

But again, please open an issue on the documentation page for documenting this better, that *will* solve your issue, because we do support playing audio like that

**Comment by Mickeon** [comment]
There may be an issue with the audio you are playing or your scene. The `max_polyphony` property was added specifically for this purpose, and for most users (me included) it works just as you would expect.

**Comment by amarraff** [comment]
@AThousandShips Yes, I read the page before I even came here. As I have said multiple times, I do not understand how to use it. So I am proposing an alternate solution for the engine that is less complicated to understand.  If you could shed any insight on how to use it, even a small code block of syntax, it could solve my problem and I would appreciate it.

@Mickeon Is there anything I need to do other than raising the value of the property? I don't understand. Like many things in the Godot documentation, the AudioStreamPlayer's page *and* the AudioStreamPolyphonic say what it is, what it can be used for, but not how to actually use it. Could you provide an example of syntax, or the steps you need to do in order to make it function? I am not convinced it is a better solution than the proposal I have submitted here.

**Comment by AThousandShips** [comment]
Just using `max_polyphony` and calling `play` on the player works exactly as you need, it's trivial

Either you are having a bug, or you are doing something wrong in your code, it doesn't require anything complicated, it just works, just increase the value and call `play` on the player and it works, it plays multiple clips at once

**Comment by AThousandShips** [comment]
So please open a bug report with clear steps for reproduction if you can't make this work with just increasing this value and calling `play`, because it works fine for me

**Comment by amarraff** [comment]
Perhaps it is a bug, because I have Max Polyphony set to 100 and the sounds are still cutting each other off as they play. Were you able to replicate this now, or are you speaking out of memory? Perhaps there is still confusion on the problem I'm having vs. the functionality you're proposing?

**Comment by AThousandShips** [comment]
No, I literally did this right now, you are having either a bug or doing something wrong 🙂

I am not confused 🙃

**Comment by Mickeon** [comment]
Keep in mind that an **AudioStreamPlayer** can only play a single stream. Every time you change `stream` all of the sound coming from the **AudioStreamPlayer** **will** stop, regardless of how many are playing at the time with `max_polyphony`.

The doubts about the documentation give me credence that it would be better to have https://github.com/godotengine/godot/pull/81858 sooner than later, I suppose

**Comment by AThousandShips** [comment]
[SoundDisplay.zip](https://github.com/godotengine/godot-proposals/files/13977292/SoundDisplay.zip)

Add a sound file to this stream player, long enough, or shorten the timeout on the timer, and you'll see

I can't send a sound file because I don't have anything open source at hand right now

**Comment by amarraff** [comment]
@AThousandShips No worries, I am confused lol. I'll give that a shot, one moment.

@Mickeon Right, and I think that's where there may be some confusion as to what I'm proposing. I'm propising to allow AudioStreamPlayers to play more than one stream at once. I am confused that you are both saying that AudioStreamPlayers can play more than one sound at once with polyphony, but then you are saying that it can't because it will only play a single stream? I am so lost.

**Comment by AThousandShips** [comment]
And that's how you use `AudioStreamPolyphonic`, that allows that

You never said you needed to play multiple streams, that should have been made clear...

But as I've already said, this should be documented better, so please open an issue about it for documentation to track this 🙂

**Comment by AThousandShips** [comment]
For the polyphonic try:
```gdscript
$MyAudioStreamPlayer.get_stream_playback().play_stream(my_stream)
```

**Comment by amarraff** [comment]
@AThousandShips 

I didn't realize that I wasn't clear with my language, sorry. I will try updating the post to have clearer language. Did you read the documentation for Unity's PlayOneShot that I provided? Perhaps that will give you a clearer understanding of what I'm proposing. I didn't realize playing multiple sounds with one stream was even possible.

I will open an issue for documentation if you would like, where do I go for that?

To your suggestion, I am not expecting you to know how everything works. You are suggesting I use a solution that I do not understand how to use, which is why I keep asking how to use it. I appreciate you giving me some code to try, it did not work. The function play_stream does not exist for get_stream_playback(). Is this perhaps something that works in Godot 4.3 but not my version? I noticed you are using Godot 4.3. I am using v4.2.1.stable.official [b09f793f5]

**Comment by AThousandShips** [comment]
It does work, I just tested it 🙃 I can send you an example project too but then I really do have to go, this works

I already linked you to the documentation above 🙂

**Comment by amarraff** [comment]
I am aware that you have linked me to the documentation, I appreciate your attempt to help. Please do not feel like you have to be the person to solve this problem for me. You are commenting, so I am replying.

Again, I am not using Godot 4.3

**Comment by Mickeon** [comment]
I feel like this entire proposal is extremely similar to https://github.com/godotengine/godot-proposals/issues/4569. Looking at the linked Unity API, it would solve the same identical issue.

**Comment by AThousandShips** [comment]
Duplicate of https://github.com/godotengine/godot-proposals/issues/4569

**Comment by amarraff** [comment]
@Mickeon Hmmm, it does seem similar, but I'm not sure if it solves the problem exactly. They suggest using a path directly to the audio. I would rather pass through an exported variant. It also does not solve the problem of using custom settings for an AudioStreamPlayer.

@AThousandShips Why did you close the issue? It is not a direct duplicate of #4569. I do not appreciate this at all. My suggestion of solution is different as I mention an inspector property as well.

**Comment by amarraff** [comment]
As I mentioned in my reply to Mickeon, that post does not address the problem of using custom AudioStreamPlayer settings to play a sound. It only instantiates a new one and then frees it. I already addressed this in my original post. You are misunderstanding the point of my post, and now it is closed so nobody else will look at it. Great.

I'm not suggesting that it's "law," kindly stop being condescending.

**Comment by amarraff** [comment]
This is the Godot Proposals forum. I proposed that we solve a solution in a way that is different from the other post. I also addressed important issues that the other post did not address, again, such as instantiating a blank AudioStreamPlayer not inheriting the settings of one pre-set in the inspector. You are misunderstanding my post, and have closed it prematurely.

You did not make the thing I need work, I already told you that the syntax you provided me did not work. The project you provided did not provide a solution. The syntax you provided for playing audio polyphonically did not work. I have stated this already.



> This already works 🙂 the proposed solution isn't needed
I don't appreciate that you closed my issue because you feel this way.

**Comment by AThousandShips** [comment]
> I don't appreciate that you closed my post because you feel this way. The enhancements I suggest should still be considered.

I didn't close it because of that, don't tell my why I did things 🙂

> You did not make the thing I need work, I already told you that the syntax you provided me did not work. The project you provided did not provide a solution. The syntax you provided for playing audio polyphonically did not work. I have stated this already.

Then open a bug report because it *does* work, I've tested it, so you're either not using it right, or there's a bug, as I've said multiple times

I have not closed this prematurely, add your ideas to the other proposal, it's sufficiently similar, we don't keep very similar proposals open separately, instead we discuss it in one place, if your solution is better than the one proposed there that one will be adjusted instead, keeping two extremely similar issues open just makes it hard to follow

Now if you want these changes and these things to work I suggest you instead add your support and ideas to that proposal, and open that bug report 🙂

**Comment by AThousandShips** [comment]
It's pretty disappointing when having spent a long time providing you with examples and instructions to have you accuse me of things I haven't done and then block me, some gratitude over the help I provided would be in place rather than that attitude you are giving instead...

In the future please consider the [code of conduct](https://godotengine.org/code-of-conduct)

**Comment by amarraff** [comment]
> I didn't close it because of that, don't tell my why I did things 🙂
I'm not. This seems to be why you closed it from my perspective... 


> Then open a bug report because it _does_ work, I've tested it, so you're either not using it right, or there's a bug, as I've said multiple times, in what way didn't it work? Did you add the sound files as instructed?
Yes. 



> It's pretty disappointing when having spent a long time providing you with examples and instructions to have you accuse me of things I haven't done and then block me, some gratitude over the help I provided would be in place rather than that attitude you are giving instead...
> 
> In the future please consider the [code of conduct](https://godotengine.org/code-of-conduct)

It's pretty disappointing when I spend an hour carefully constructing a post that I felt passionately about, only to have it closed by someone who personally deems it as an irrelevant duplicate. I'm sure you all deal with plenty of nonsense posts, but I am genuinely trying to help improve the engine.

I am grateful for your attempt to help me, I already said that.

**Comment by AThousandShips** [comment]
> I'm not. This seems to be why you closed it from my perspective...

But you said I did close it because of that, that's an accusation...

You feeling upset that I said this is a duplicate because it covers the same things and that you should just add your ideas to the other post isn't a reason to behave in this way, and making your responses emotional and accusing me of things, or becoming non constructive

Please don't take closing a post as personal, or that it's irrelevant, and I understand that you're trying to make the engine better, but you don't do that by focusing on that *your* proposal gets to be open, but by focusing on one thing, as I said, it's not helpful to have a bunch of similar proposals open with a bunch of people commenting on each, that's why it's closed, not because it's irrelevant or because I dislike it

I can call for some other maintainer to make the decision of being a duplicate if you want

But please act professional and don't take your proposal being closed as a personal attack, or dismiss the statement that the feature you seem to be asking for, playing multiple streams uninterrupted, is already well supported 🙂

**Comment by amarraff** [comment]
> But you said I did close it because of that, that's an accusation...

No, it wasn't. You can take it that way if you please. It genuinely seemed to me that's why you did it. 

> You feeling upset that I said this is a duplicate because it covers the same things and that you should just add your ideas to the other post isn't a reason to behave in this way, and making your responses emotional and accusing me of things, or becoming non constructive
> 
> Please don't take closing a post as personal, or that it's irrelevant, and I understand that you're trying to make the engine better, but you don't do that by focusing on that _your_ proposal gets to be open, but by focusing on one thing, as I said, it's not helpful to have a bunch of similar proposals open with a bunch of people commenting on each, that's why it's closed, not because it's irrelevant or because I dislike it, you must surely see how it's reasonable to not have multiple proposals open at once like that

I do. As I said, I believe my post is different enough to keep it open. I understand that you disagree. My goal is to provide a solution to this problem that is more user-friendly, similar to how Unity accomplishes it. The other post did not discuss that.

> I can call for some other maintainer to make the decision of being a duplicate if you want

No, that's fine. I imagine they'll come to the same conclusion based on the reasoning you've provided.

> But please act professional and don't take your proposal being closed as a personal attack, or dismiss the statement that the feature you seem to be asking for, playing multiple streams uninterrupted, is already well supported 🙂

I am voicing my displeasure and disagreeing with you, I don't see how that is unprofessional. Disagreements are natural. I'm aware that it's not a personal attack, but that doesn't make it any less frustrating.  I tried out both projects you provided me, I have tried implementing your code from the project you provided me, and it is not working. I will keep trying.

Have a nice day.

**Comment by AThousandShips** [comment]
Then you can either add your ideas to the other proposal and help improve it, or not

And I'm sorry if it's frustrating to have spent effort on something that ends up being closed, but that's not a reason to keep it open, nor a reason to behave this way

And I am not going to accept you dismissing the fact that you accused me of closing it for reasons you didn't, I'd ask that you apologize, because this is what you said:
> you closed my post because you feel this way

Don't gaslight me...

**Comment by amarraff** [comment]
> Then you can either add your ideas to the other proposal and help improve it, or not
> 
> And I'm sorry if it's frustrating to have spent effort on something that ends up being closed, but that's not a reason to keep it open, nor a reason to behave this way

You're misconstruing why I'm frustrated, it is not solely because it ended up being closed.

As I already said, I am voicing my displeasure and disagreeing with you, which I am allowed to do. I'm sorry if you do not like what I have to say. I do not appreciate how you handled this post. I'm not sure what you mean by me "behaving this way."

> Don't gaslight me...

I am not, and now you are accusing me of gaslighting, which I do not appreciate. I have edited the post in question so you do not feel as though I am trying to gaslight you.

Have a nice day.

**Comment by AThousandShips** [comment]
You editing what you said instead of apologizing for what you did say is gaslighting...

That's it, I'm done and reporting this

**Comment by amarraff** [comment]
> You editing what you said instead of apologizing for what you did say is gaslighting...
> 
> That's it, I'm done and reporting this

I've made it pretty clear that that wasn't my intention. You very clearly just want to make up a reason to report me. I see through what you're doing here.

**Comment by AThousandShips** [comment]
I'm not making anything up, you are just accusing me again of things I have not done...

CC @akien-mga @hpvb sorry to bother you about this

**Comment by amarraff** [comment]
@AThousandShips I was just discussing our conversation with my partner; I feel bad that I came across in a way I didn’t intend, I chose my words poorly due to my confusion and frustration with not understanding this aspect of Godot. So I want to apologize. I sincerely did not mean to come across like I was trying to gaslight you, and when I told you I edited my post I was genuinely trying to admit fault with my wording, not upset you. I also should not have given into my defensive impulse and blocked you. I really appreciate all of your help. Thank you for helping me understand this, I’m sorry.

**Comment by AThousandShips** [comment]
Thank you 🙂

**Comment by YuriSizov** [comment]
Hey, @amarraff, I just wanted to clarify that when proposals or issues are closed as a duplicate this is not because they are not valuable on their own. He reality is we have thousands of conversations going on, so we try to consolidate or resolve as many as we can, to make it easier for everyone participating to find the current status and to reach the resolution.

Please try not to take it personally that your suggestion was closed as a duplicate. You can still continue the discussion when it happens, and your ideas are still appreciated and should be considered.
